File: Ai algorithm implementation/First-Choice_Hill_Climbing_(N-Queens).cpp
#include <bits/stdc++.h>
using namespace std;

int N = 8;

int heuristic(vector<int> &state) {
    int h = 0;
    for(int i = 0; i < N; i++)
        for(int j = i + 1; j < N; j++)
            if(state[i] == state[j] || abs(state[i] - state[j]) == abs(i - j))
                h++;
    return h;
}

vector<int> firstChoiceHC(vector<int> state, int maxTrials = 50) {
    while (true) {
        bool improved = false;
        int hCurrent = heuristic(state);

        for (int t = 0; t < maxTrials; t++) {
            int col = rand() % N;
            int row = rand() % N;
            if (row == state[col]) continue;

            vector<int> newState = state;
            newState[col] = row;

            int hNew = heuristic(newState);
            if (hNew < hCurrent) {
                state = newState;
                improved = true;
                break;
            }
        }

        if (!improved) break;
    }
    return state;
}

int main() {
    srand(time(0));
    vector<int> state(N);
    for (int i = 0; i < N; i++)
        state[i] = rand() % N;

    vector<int> solution = firstChoiceHC(state);
    cout << "First-Choice Hill Climbing Solution (Heuristic=" << heuristic(solution) << "):\n";
    for (int i = 0; i < N; i++)
        cout << solution[i] << " ";
    cout << "\n";
}

File: Ai algorithm implementation/First-Choice_Hill_Climbing_(N-Queens).py
import random

N = 8

def heuristic(state):
    h = 0
    for i in range(N):
        for j in range(i + 1, N):
            if state[i] == state[j] or abs(state[i] - state[j]) == abs(i - j):
                h += 1
    return h

def first_choice_hc(state, max_trials=50):
    while True:
        improved = False
        h_current = heuristic(state)
        
        for t in range(max_trials):
            col = random.randint(0, N - 1)
            row = random.randint(0, N - 1)
            if row == state[col]:
                continue
            
            new_state = state[:]
            new_state[col] = row
            
            h_new = heuristic(new_state)
            if h_new < h_current:
                state = new_state
                improved = True
                break
        
        if not improved:
            break
    
    return state

random.seed()
state = [random.randint(0, N - 1) for _ in range(N)]

solution = first_choice_hc(state)
print(f"First-Choice Hill Climbing Solution (Heuristic={heuristic(solution)}):")
print(solution)
